dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/Program.java

```diff
@@ -904,4 +915,67 @@ public void setAccessLevel( AccessLevel accessLevel )
         this.accessLevel = accessLevel;
     }
+
+    public static final BiFunction<Program, Map<String, String>, ProgramCopyTuple> copyOf = ( original, options ) -> {
+        Program copy = new Program();
+        copy.setAutoFields();
+        setShallowCopyValues( copy, original, options );
+        setDeepCopyValues( copy, original, options );
+        return new ProgramCopyTuple( copy, original );
+    };
+
+    private static void setShallowCopyValues( Program copy, Program original, Map<String, String> options )
+    {
+        String prefix = options.getOrDefault( PREFIX_KEY, DEFAULT_PREFIX );
+        copy.setAccessLevel( original.getAccessLevel() );
+        copy.setCompleteEventsExpiryDays( original.getCompleteEventsExpiryDays() );
+        copy.setDescription( original.getDescription() );
+        copy.setDisplayIncidentDate( original.getDisplayIncidentDate() );
+        copy.setDisplayFrontPageList( original.getDisplayFrontPageList() );
+        copy.setEnrollmentDateLabel( original.getEnrollmentDateLabel() );
+        copy.setExpiryDays( original.getExpiryDays() );
+        copy.setFeatureType( original.getFeatureType() );
+        copy.setFormName( original.getFormName() );
+        copy.setIgnoreOverdueEvents( original.getIgnoreOverdueEvents() );
+        copy.setIncidentDateLabel( original.getIncidentDateLabel() );
+        copy.setMaxTeiCountToReturn( original.getMaxTeiCountToReturn() );
+        copy.setMinAttributesRequiredToSearch( original.getMinAttributesRequiredToSearch() );
+        copy.setName( prefix + original.getName() );
+        copy.setOnlyEnrollOnce( original.getOnlyEnrollOnce() );
+        copy.setOpenDaysAfterCoEndDate( original.getOpenDaysAfterCoEndDate() );
+        copy.setProgramType( original.getProgramType() );
+        copy.setPublicAccess( original.getPublicAccess() );
+        copy.setSharing( original.getSharing() );
+        copy.setShortName( original.getShortName() );
+        copy.setSelectEnrollmentDatesInFuture( original.getSelectEnrollmentDatesInFuture() );
+        copy.setSelectIncidentDatesInFuture( original.getSelectIncidentDatesInFuture() );
+        copy.setSkipOffline( original.isSkipOffline() );
+        copy.setUseFirstStageDuringRegistration( original.getUseFirstStageDuringRegistration() );
+        copy.setCategoryCombo( original.getCategoryCombo() );
+        copy.setDataEntryForm( original.getDataEntryForm() );
+        copy.setExpiryPeriodType( original.getExpiryPeriodType() );
+        copy.setNotificationTemplates( newSetFromObjectOrEmpty( original.getNotificationTemplates() ) );
+        copy.setOrganisationUnits( newSetFromObjectOrEmpty( original.getOrganisationUnits() ) );
+        copy.setProgramAttributes( newListFromObjectOrEmpty( original.getProgramAttributes() ) );
+        copy.setProgramIndicators( newSetFromObjectOrEmpty( original.getProgramIndicators() ) );
+        copy.setProgramRuleVariables( newSetFromObjectOrEmpty( original.getProgramRuleVariables() ) );
```

Comment:
I think we need to create new entities for those `ProgramRuleVariable`s because we have one-to-many assocation between `Program` and `ProgramRuleVariable`. 
Same for other properties such as `ProgramSection`, `ProgramIndicator` , `ProgramTrackedEntityAttribute`.

Reply:
Yep, I see your point. It does look like they are needed as deep copies also. I wonder was there any specific reason why these were not included in the original Jira issue. @larshelge does this sound right to you also?
So these would be the extra deep copies required:
- `Set<ProgramSection> programSections`
- `List<ProgramTrackedEntityAttribute> programAttributes`
- `Set<ProgramIndicator> programIndicators`
- `Set<ProgramRuleVariable> programRuleVariables`